scheduler/src/cook/util.clj

```diff
@@ -230,3 +231,11 @@
     (let [old-val @atom
           swap-happened (compare-and-set! atom old-val newval)]
       (if swap-happened old-val (recur)))))
+
+(defn log-structured
+  "Log a structured message at the info level.
+  Takes as input a metadata dictionary and a data string.
+  Can specify a custom logger namespace if desired.
+  "
+  ([data metadata] (log/info (json/write-str (assoc metadata :data data))))
```

Comment:
I think having all structured logs in a separate logger namepsace is not exactly what we want, since we also intend to use this functionality for adding pool/k8s cluster fields to other metrics, eg in lines like this one https://github.com/twosigma/Cook/blob/master/scheduler/src/cook/scheduler/scheduler.clj so we can avoid having to say "in <pool>"...

But then we *also* want to use this for the match cycle metrics, which Laura + Lewis suggested should go to a separate logstash index to improve query performance, and our idea was to use the unique logger namespace to split them out in our logstash config.

So, I think what I'll do is define the macro but still leave the optional logger-ns param for the specific match cycle metrics.

Reply:
Okay I went back to it and I think I initially misunderstood your comment, but I think I understand now. I did define a separate namespace for the structured logging, and used `defmacro`'s so that it preserves the logger namespace of the calling context. I did keep the optional `logger-ns` parameter (moved to the end of the function) so that we can have the separate index for match cycle metrics.